boards/sim/sim/sim/scripts/Make.defs

```diff
@@ -31,6 +31,10 @@ ifneq ($(CONFIG_DEBUG_NOOPT),y)
   ARCHOPTIMIZATION += -O2 -fno-strict-aliasing
 endif
 
+ifeq ($(CONFIG_STACK_CANARIES),y)
+  ARCHOPTIMIZATION += -fstack-protector-all
+endif
```

Comment:
what's the intended scope of this change?

ARCHOPTIMIZATION seems a bit too ambitious as it would affect CELFFLAGS, CMODULEFLAGS, and even HOSTCFLAGS.

Reply:
Yes, the stack protection mechanism don't have the external dependence, so it work in all these environment.

Reply:
i'm not sure if i get you intention.

* do you mean to make HOSTSRCS to use nuttx ssp things (libs/libc/stdlib/lib_stackchk.c) instead of the host provided ones? it sounds awkward to me.

* how are those symbols resolved for MODULES and ELF?

Reply:
> i'm not sure if i get you intention.
> 
> * do you mean to make HOSTSRCS to use nuttx ssp things (libs/libc/stdlib/lib_stackchk.c) instead of the host provided ones? it sounds awkward to me.

Yes, both HOSTSRCS and SRCS use libs/libc/stdlib/lib_stackchk.c. We can't provide two implementation here because compiler hardcode the name of cookie and funntion.  And gcc don't link toolchain's libssp.a by default. it's just like NuttX replace toolchain's libc/libm.

> * how are those symbols resolved for MODULES and ELF?

See as other normal libc function.
SSP is almost same as the old mechanism but very fast:
https://github.com/apache/incubator-nuttx/blob/master/arch/arm/src/armv7-m/arm_stackcheck.c
and can be implemented in the arch independent way.

Reply:
@xiaoxiang781216 

Just an FYI with armv7-m/arm_stackcheck.c the code size blows up and performance drops considerably, there is also a  reserved register that needs initialization. What are the effect of SSP in these terms? Do you have any metrics?

Reply:
> Just an FYI with armv7-m/arm_stackcheck.c the code size blows up and performance drops considerably, there is also a reserved register that needs initialization. What are the effect of SSP in these terms? Do you have any metrics?

No, SSP don't need resolve a special register. Here is the data without SSP:
```
   text    data     bss     dec     hex filename
  80900     320    3572   84792   14b38 nuttx
```
and with SSP:
```
   text    data     bss     dec     hex filename
 107092     320    3572  110984   1b188 nuttx
```

Reply:
So a 33% increase. The armv7 Runtime chaedck one was on a function call entry as well, but it was a margin guess that can be overly aggressive and tending to waste some memory. See https://cwiki.apache.org/confluence/display/NUTTX/ARMv7-M+Run+Time+Stack+Checking

I like this! I would keep both!

Reply:
Yes, you can select both from Kconfig, I intention to make both work indepently or you can enable both at the same time. Normally, -fstack-protector-strong is eough to catch all type of error which just increase (85624 - 81604)/81604=5%.